### PR TITLE
Implement attribute restyle hints

### PR DESCRIPTION
**Pull request by bholley**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8381)

### CODE REVIEW

**Comment by nox** [comment]
What is the rationale for moving AttrValue in style exactly?

File: components/script/dom/attr.rs [hunk 1]
```diff
@@ -181,11 +36,13 @@ impl Attr {
                      prefix: Option<Atom>, owner: Option<&Element>) -> Attr {
         Attr {
             reflector_: Reflector::new(),
-            local_name: local_name,
+            name: AttrName {
+                local_name: local_name,
+                name: name,
+                namespace: namespace,
+                prefix: prefix,
+            },
```
Comment (nox) [inline components/script/dom/attr.rs]: Could you name this field and structure differently? `foo.foo` also sounded silly to me. Maybe `AttrID`, or `AttrIdentity`, or `AttrInfo`, or something entirely different that I can't think of?

File: components/script/dom/attr.rs [hunk 2]
```diff
@@ -291,33 +148,41 @@ impl AttrMethods for Attr {
 impl Attr {
     pub fn set_value(&self, mut value: AttrValue, owner: &Element) {
         assert!(Some(owner) == self.owner().r());
+        let notify = self.name.namespace == ns!("");
+        if notify {
+            owner.will_mutate_attr();
```
Comment (nox) [inline components/script/dom/attr.rs]: Shouldn't that be called for all namespaces, given CSS attributes selectors aren't limited to the empty namespace?

**Comment by bholley** [comment]
> What is the rationale for moving AttrValue in style exactly?
Because restyle_hints.rs needs to use it, and |style| can't depend on |script|.

File: components/script/dom/attr.rs (hunk 1 above)
Comment (bholley) [inline components/script/dom/attr.rs]: I can switch it to AttrIdentifier if you like.

File: components/script/dom/attr.rs (hunk 2 above)
Comment (bholley) [inline components/script/dom/attr.rs]: Probably, yeah. I was trying to mimic the existing behavior here, but the existing behavior is probably just wrong.

**Comment by jdm** [comment]
@bors-servo: retry
#8416

**Comment by pcwalton** [comment]
Looks good modulo a few nits

**Comment by Manishearth** [comment]
```
./components/style/restyle_hints.rs:292: Line is longer than 120 characters
```

**Pull request merged**